optuna/pruners/threshold.py

```diff
@@ -0,0 +1,125 @@
+import math
+from typing import Optional
+
+import optuna
+from optuna.pruners import BasePruner
+from optuna.pruners.percentile import _is_first_in_interval_step
+
+
+class ThresholdPruner(BasePruner):
+    """Pruner to detect outlying metrics of the trials.
+
+    Prune if a metric exceeds upper bound threshold or
+    falls behind lower bound threshold.
+
+    Example:
+        .. testcode::
+
+            from optuna import create_study
+            from optuna.exceptions import TrialPruned
+            from optuna.pruners import ThresholdPruner
+
+
+            def objective_for_upper(trial):
+                for step, y in enumerate(ys_for_upper):
+                    trial.report(y, step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+                return ys_for_upper[-1]
+
+
+            def objective_for_lower(trial):
+                for step, y in enumerate(ys_for_lower):
+                    trial.report(y, step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+                return ys_for_lower[-1]
+
+
+            ys_for_upper = [0.0, 0.1, 0.2, 0.5, 1.2]
+            ys_for_lower = [100.0, 90.0, 0.1, 0.0, -1]
+            n_trial_step = 5
+
+            study = create_study(pruner=ThresholdPruner(upper=1.0))
+            study.optimize(objective_for_upper, n_trials=10)
+
+            study = create_study(pruner=ThresholdPruner(lower=0.0))
+            study.optimize(objective_for_lower, n_trials=10)
+
+
+    Args
+        lower:
+            A minimum value which determines whether pruner prunes or not.
+            If an intermediate value is smaller than lower, it prunes.
+        upper:
+            A maximum value which determines whether pruner prunes or not.
+            If an intermediate value is larger than upper, it prunes.
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+        self,
+        lower: Optional[float] = None,
+        upper: Optional[float] = None,
+        n_warmup_steps: int = 0,
+        interval_steps: int = 1,
+    ) -> None:
+
+        if isinstance(lower, bool) or isinstance(upper, bool):
+            raise ValueError("lower and upper should be either None, integers or floating points")
```

Comment:
Thank you for the suggestion, I applied your suggestion (partially). b516491b

> Does the bool type need a separate check?

I check the `bool` separately since `isinstance(True, int)` returns `True`.
I'm not sure if it's enough to check `(int, float)`.
What do you think?

Reply:
b516491 :bow: